### PR TITLE
Add reorg cache to manager as well as a fast catch up mode

### PR DESCRIPTION
**Pull request by hkalodner**
- Adds a fast catch up mode to the validator which speeds up processing of blocks older than the reorg threshold
- During reorg, handle calls to validator from a cache of the most recent pre-reorg state rather than blocking
- On initial boot of validator respond to calls with an error rather than blocking

### CODE REVIEW

File: packages/arb-validator/rollupmanager/manager.go [hunk 1]
```diff
@@ -38,20 +40,23 @@ const (
...
-	// an appropriate way to resolve them
-	validCallLock sync.Mutex
+type Manager struct {
+	// The mutex should be held whenever listeres or reorgCache are accessed or
+	// set and whenever activeChain is going to be set (but not when it is accessed)
+	sync.Mutex
 
 	listeners   []rollup.ChainListener
 	activeChain *rollup.ChainObserver
+	reorgCache  *reorgCache
```
Comment (edfelten) [inline packages/arb-validator/rollupmanager/manager.go:59]: Should add a comment explaining what this is for and the limit on how it should be used.

File: packages/arb-validator/rollupmanager/manager.go [hunk 2]
```diff
@@ -146,77 +150,132 @@ func CreateManagerAdvanced(
...
-				}
+					distanceFromCurrent := new(big.Int).Sub(currentL1Height, currentLocalHeight)
+					if distanceFromCurrent.Cmp(maxReorg) <= 0 {
+						log.Println("Ending fast catchup")
+						break
+					}
 
-				man.activeChain.NotifyNewBlock(blockId.Clone())
-				log.Print(man.activeChain.DebugString("== "))
+					fetchEnd := new(big.Int).Add(currentLocalHeight, maxReorg)
```
[the remark this answers is not in the document]
Comment (hkalodner) [reply packages/arb-validator/rollupmanager/manager.go:174]: Ah good catch. That was my intention, but not the way it's implemented

File: packages/arb-validator-core/arbbridge/events.go [hunk 3]
```diff
@@ -28,6 +28,35 @@ type Event interface {
 	GetChainInfo() ChainInfo
 }
 
+// MergeEventsUnsafe assumes that both sets of events come from the same
+// chain state rather than from two different states caused by a reorg
+func MergeEventsUnsafe(events1 []Event, events2 []Event) []Event {
```
[the remark this answers is not in the document]
Comment (hkalodner) [reply packages/arb-validator-core/arbbridge/events.go:33]: I'll add a comment saying that it also expects disjointness. I could also move this function into the `manager.go` file, but it seemed like there was so chance that we might want to use it elsewhere

File: packages/arb-validator/rollupmanager/manager.go (hunk 1 above)
Comment (hkalodner) [reply packages/arb-validator/rollupmanager/manager.go]: I added a lengthy comment

File: packages/arb-validator/rollupmanager/manager.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hkalodner) [reply packages/arb-validator/rollupmanager/manager.go]: Updated behavior and added a comment describing its behavior

File: packages/arb-validator-core/arbbridge/events.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hkalodner) [reply packages/arb-validator-core/arbbridge/events.go:33]: Documented that the event slices are assumed to be disjoint

**Pull request merged**